aten/src/ATen/core/jit_type_base.h

```diff
@@ -137,6 +137,11 @@ struct TORCH_API Type : std::enable_shared_from_this<Type> {
     return false;
   }
 
+  // `OptionalType` has been deprecated and replaced by `UnionType`.
+  // This is a convenience method so that we don't have to replace
+  // all `== OptionalType::Kind` checks with a cast and another method
+  bool isOptional() const;
```

Comment:
Making it a static method on UnionType (that is, `UnionType::isOptional(TypePtr type)` means that we can't pass in `this` as an argument for `type`.

On a related note, I also had problems returning `c10::optional<UnionTypePtr>` instead of `bool`. The method has to be `const` because of where it's used in the codebase, but calling `Type::cast<const T>` returns a `shared_ptr<const T>`. The only way I could get around this and return a `UnionTypePtr` is by making an entirely new `UnionType`, which is obviously wasteful.

Please let me know what you think the best course of action would be